Docker.DotNet/Microsoft.Net.Http.Client/BufferedReadStream.cs

```diff
@@ -0,0 +1,242 @@
+        {
+            get { return _inner.CanRead || _bufferCount > 0; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override bool CanTimeout
+        {
+            get { return _inner.CanTimeout; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return _inner.CanWrite; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotSupportedException(); }
+        }
+
+        public override long Position
+        {
+            get { throw new NotSupportedException(); }
+            set { throw new NotSupportedException(); }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _disposed = true;
+            if (disposing)
+            {
+                _inner.Dispose();
+            }
+        }
+
+        public override void Flush()
+        {
+            _inner.Flush();
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            return _inner.FlushAsync(cancellationToken);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            _inner.Write(buffer, offset, count);
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return _inner.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            // Validate Inputs
+
+            // Drain buffer
+            if (_bufferCount > 0)
+            {
+                int toCopy = Math.Min(_bufferCount, count);
+                Buffer.BlockCopy(_buffer, _bufferOffset, buffer, offset, toCopy);
+                _bufferOffset += toCopy;
+                _bufferCount -= toCopy;
+                return toCopy;
+            }
+
+            return _inner.Read(buffer, offset, count);
+        }
+
+        public async override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // Validate Inputs
+
+            // Drain buffer
+            if (_bufferCount > 0)
+            {
+                int toCopy = Math.Min(_bufferCount, count);
+                Buffer.BlockCopy(_buffer, _bufferOffset, buffer, offset, toCopy);
+                _bufferOffset += toCopy;
+                _bufferCount -= toCopy;
+                return toCopy;
+            }
+
+            return await _inner.ReadAsync(buffer, offset, count, cancellationToken);
+        }
+
+        private void EnsureBufferd()
+        {
```

Comment:
Also for the sync async pairs a common pattern is to call

```C#
var t = this.EnsureBufferedAsync().ConfigureAwait(false);
t.Wait
```
instead of duplicating the code.